### PR TITLE
Make TrajectoryConnectivity use trajectory IDs.

### PR DESCRIPTION
**Pull request by SirVer**
Based on #274 and related to #256.

### CODE REVIEW

**Comment by SirVer** [comment]
Finally a net negative 8).

File: cartographer/mapping_2d/sparse_pose_graph.h [hunk 1]
```diff
@@ -112,8 +111,8 @@ class SparsePoseGraph : public mapping::SparsePoseGraph {
     // which are finished.
     bool finished = false;
 
-    // The trajectory to which this SubmapState belongs.
-    const mapping::Submaps* trajectory = nullptr;
+    // The ID of the trajectory to which this SubmapState belongs.
+    int trajectory_id = -1;
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph.h]: Remove since this is redundant now?

File: cartographer/mapping_3d/sparse_pose_graph.cc [hunk 2]
```diff
@@ -106,16 +107,16 @@ void SparsePoseGraph::AddScan(
...
-  trajectory_connectivity_.Add(trajectory);
+  trajectory_connectivity_.Add(trajectory_id);
 
   if (submap_indices_.count(insertion_submaps.back()) == 0) {
     submap_indices_.emplace(insertion_submaps.back(),
                             static_cast<int>(submap_indices_.size()));
     submap_states_.emplace_back();
     submap_states_.back().submap = insertion_submaps.back();
-    submap_states_.back().trajectory = trajectory;
+    submap_states_.back().trajectory_id = trajectory_id;
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph.cc]: Similar to 2D: Remove the member from the struct. Similar for other 2D related comments.

File: cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc [hunk 3]
```diff
@@ -227,7 +224,10 @@ void ConstraintBuilder::ComputeConstraint(
             initial_pose.rotation(), filtered_point_cloud, point_cloud,
             options_.global_localization_min_score(), &score, &pose_estimate)) {
       CHECK_GT(score, options_.global_localization_min_score());
-      trajectory_connectivity->Connect(scan_trajectory, submap_trajectory);
+      CHECK_GE(0, scan_trajectory_id);
```
Comment (wohe) [inline cartographer/mapping_3d/sparse_pose_graph/constraint_builder.cc]: `CHECK_GE(scan_trajectory_id, 0)`, right? Similar on the next line.

File: cartographer/mapping_2d/sparse_pose_graph/constraint_builder.cc [hunk 4]
```diff
@@ -194,7 +191,8 @@ void ConstraintBuilder::ComputeConstraint(
             filtered_point_cloud, options_.global_localization_min_score(),
             &score, &pose_estimate)) {
       CHECK_GT(score, options_.global_localization_min_score());
-      trajectory_connectivity->Connect(scan_trajectory, submap_trajectory);
+      trajectory_connectivity->Connect(scan_trajectory_id,
```
Comment (wohe) [inline cartographer/mapping_2d/sparse_pose_graph/constraint_builder.cc]: Add the CHECK from 3D here as well?

**Pull request merged**